applications/frontend/src/middleware/backend.ts

```diff
@@ -0,0 +1,59 @@
+import { MiddlewareAPI, Dispatch, Middleware, AnyAction } from "redux";
+
+export type CallApiPayload = {
+  successAction: string;
+  errorAction: string;
+  url: RequestInfo;
+  params?: RequestInit;
+}
+
+export type CallApiAction = {
+  type: string;
+  payload: {
+    successAction: string;
+    errorAction: string;
+    url: RequestInfo;
+    params?: RequestInit;
+  };
+  meta: {
+    callApi: true;
+  };
+};
+
+/**
+ * @private
+ */
+function createCallAPIMiddleware(
+  fetchFn: (input: RequestInfo, init?: RequestInit) => Promise<Response>
+) {
+  const callAPIMiddleware: Middleware<Dispatch> = ({
+    dispatch
+  }: MiddlewareAPI) => next => (action: AnyAction | CallApiAction) => {
+    if (!action.meta || !action.meta.callApi) {
```

Comment:
We can do better here with a further level of abstraction. Let's say that I have a search going to the backend. The "search_action" is dispatched by the component, unaware of the api. The middleware intercepts the "SEARCH_ACTION", calls the backend and when finished continues with the action "SEARCH_PERFORMED" (`next('SEARCH_PERFORMED') with the result in the payload (or whatever is needed to retrieve it if we don't want to put all the data in the redux store). What we obtain with this is that becomes very easy to change to a mocked backend directly returning the data, leaving all the details about api communication inside the middleware.

Reply:
that's what this middleware does if the action has meta callApi then it does the fetch and on fetch success dispatch the success action with the payload